File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        is_rotated = nums[0] > nums[-1]
        if not is_rotated:
            return nums[0]
        
        # find rotation point: find the first element that is less than nums[0]
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            if nums[mid] >= nums[0]:
                left = mid + 1
            else:
                right = mid

        return nums[left]

File: Data Structures & Algorithms/sqrtx/submission-1.cs
public class Solution {
    public int MySqrt(int x) {
        if (x < 2)
        {
            return x;
        }

        int left = 0, right = x/2;
        
        while (left < right)
        {
            var mid = left + (right - left) / 2 + 1;
            if (mid > x / mid)
            {
                right = mid - 1;
            }
            else
            {
                left = mid;
            }
        }

        return left;
    }
}